### ISSUE TITLE
bug: `programs.tmux` options can override tmux-sensible, `/etc/tmux.conf` inadvertently

### ISSUE THREAD

**Issue by lilyball** [nix-community/home-manager#2541]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Issue description

`programs.tmux` has various options such as `terminal`, `aggressiveResize`, `historyLimit`, etc. These default to the current default values that tmux uses, so the assumption is that if you don't touch them, it doesn't matter as they're just setting the options to the values they already have.

Unfortunately, this is not true. The easiest way to make this not true is to set `programs.tmux.sensibleOnTop = true`, which will load [tmux-sensible](https://github.com/tmux-plugins/tmux-sensible) prior to any other configuration. tmux-sensible sets a bunch of these options, and the default home-manager module configuration will then immediately reset some of them back to the original tmux defaults. But even with `sensibleOnTop` off, you can still run into this problem if you have an `/etc/tmux.conf` that sets up system-wide configuration (I believe nix-darwin's tmux support will create this file, and I assume—though haven't checked—that nixos will too).

My suggestion would be to have these options have no default at all but use `defaultText` to describe the tmux default, and then to have the actual configuration check if a value has been set. Alternatively they could allow `null` and use that as the default (as `prefix` does) or check if the value is different than the default (as `shortcut` does), but those feel like hackier solutions (especially the shortcut one; that really should just switch over to having no default and checking if the option is defined).

### Maintainer CC

@marsam @toonn @expipiplus1

### System information

```markdown
- system: `"x86_64-darwin"`
 - host os: `Darwin 21.1.0, macOS 10.16`
 - multi-user?: `yes`
 - sandbox: `relaxed`
 - version: `nix-env (Nix) 2.4`
 - channels(lily): `"darwin, home-manager"`
 - channels(root): `"nixpkgs-21.11pre329221.c11d08f0239"`
 - nixpkgs: `/etc/nix/inputs/nixpkgs`
```

**Comment by lilyball** [comment]
To clarify, I think the following configuration should produce something very minimal:

```nix
{
    programs.tmux.enabled = true;
}
```

Some of the options, such as `keyMode`, it seems reasonable to set unconditionally. That's not something that really should be set at the system level. But anything that tmux-sensible touches (or might plausibly touch in the future) and currently defaults to the tmux default should not end up in the generated configuration file if I haven't touched the value.

**Comment by toonn** [comment]
One concern here is that it's not quite backwards compatible to change this. I do think `null` should be a legal value to say "Don't mess with this setting." I also agree it's a bit unfortunate to always override with the program's defaults but as long as a way is introduced to disable the overriding behavior it's not terrible and IMO it would violate POLA if the module documentation shows defaults that don't actually apply unless you set them explicitly.

**Comment by lilyball** [comment]
It's very nearly backwards-compatible to change this when `sensibleOnTop` is off (AFAIK it only differs if `/etc/tmux.conf` exists, and if it is setting stuff that `programs.tmux` is overwriting by accident then it's probably good to fix that). With `sensibleOnTop` on, changing this is backwards-incompatible, but again, the current behavior is rather buggy in this scenario, so again it seems like a good fix to make. In theory you could deprecate `sensibleOnTop`, preserve existing behavior if it's on but otherwise fix it, and use a new variable that does the same thing without existing behavior, but that seems overly complicated.

As it is, I'm currently just dealing with this by leaving `sensibleOnTop` off and putting `sensible` as the first entry in `tmux.plugins`, but this does leave me vulnerable to ordering issues if multiple modules set plugins. The real solution is to use multiple modules explicitly to force ordering of sensible to the top while leaving the rest of the plugins at the default behavior, but that's annoying to write and so I haven't. I just hope this doesn't bite me later.

> IMO it would violate POLA if the module documentation shows defaults that don't actually apply unless you set them explicitly.

With `defaultText` you can write it to make it clear that this is the tmux default, as opposed to the actual value in Nix. Or just skip that and list the tmux default in the description.

**Comment by lilyball** [comment]
This still bothers me.

**Comment by pimvh** [comment]
Hi! I am loving home-manager so far, but I also find the default `tmux` configuration quite opinionated, which is frankly also an opinion 😄. Would you be open to a PR that wraps all the default configuration in a boolean flag that is set to `true` for backward compatibility?

**Comment by toonn** [comment]
@pimvh, this is about the tmux module. In principle you can overwrite anything in `$HOME` pretty easily with HM. The module's just too opinionated.

**Comment by pimvh** [comment]
> @pimvh, this is about the tmux module. In principle you can overwrite anything in `$HOME` pretty easily with HM. The module's just too opinionated.

Yes, indeed, they are different. For other people finding this issue my solution to get around this while using the home-manager module is doing this: 

```
    tmux = {
      enable = true;
      sensibleOnTop = false;
      extraConfig = ''
        # Either source another config file, you push as a managed home-manager file (see xdg.Configfile)
        source-file ~/.config/tmux/main.conf
        
        # our just specify your tmux config here, if you like embedding config in nix.
        .....
      '';
```

As explained in text by @lilyball above.

**Comment by daniels** [comment]
> As it is, I'm currently just dealing with this by leaving `sensibleOnTop` off and putting `sensible` as the first entry in `tmux.plugins`, but this does leave me vulnerable to ordering issues if multiple modules set plugins.

This too produces unexpected result, as `sensible` only changes defaults that hasn't been tampered with. (Which makes the setting sensibleOnTop meaningless. Sensible should have the same effect if placed first or last.) By letting the hm-defaults reset the tmux defaults explicitly in the generated config there is really no way to let sensible handle these options.

Maybe there should be a different sensible-setting - "useSensibleDefaults" that does the same thing as sensible - either by calling out to it *and* preventing the conflicting hm-manager-defaults to be set, *or* by simply changing the values set by the hm-generated config to the values from sensible plugin?

**Comment by Munksgaard** [comment]
I'm also annoyed by this.

**Comment by Munksgaard** [comment]
I use the following config:

```nix
    tmux = {
      enable = true;

      clock24 = true;
      plugins = with pkgs.tmuxPlugins; [
        {
          plugin = dracula;
          extraConfig = ''
            set -g @dracula-show-powerline true
            set -g @dracula-show-battery false
            set -g @dracula-plugins " "
          '';
        }
      ];

      sensibleOnTop = true;
    };
```

Since sensible sets a history limit of 50000, I'd hoped that would be my history limit. Unfortunately, home-manager then immediately overwrites it with the default 2000.

**Comment by lilyball** [comment]
> > As it is, I'm currently just dealing with this by leaving `sensibleOnTop` off and putting `sensible` as the first entry in `tmux.plugins`, but this does leave me vulnerable to ordering issues if multiple modules set plugins.
> 
> This too produces unexpected result, as `sensible` only changes defaults that hasn't been tampered with. (Which makes the setting sensibleOnTop meaningless. Sensible should have the same effect if placed first or last.) By letting the hm-defaults reset the tmux defaults explicitly in the generated config there is really no way to let sensible handle these options.

`sensible` defines "not tampered with" as "the option value matches the default". So any option that the hm module sets to its default value still matches its default and therefore `sensible` will happily update it.

**Comment by Munksgaard** [comment]
> `sensible` defines "not tampered with" as "the option value matches the default". So any option that the hm module sets to its default value still matches its default and therefore `sensible` will happily update it.

Not if sensible is run before those options from the hm module is set.

Currently, the resulting tmux when using `sensibleOnTop` looks something like this:

```conf
# ============================================= #
# Start with defaults from the Sensible plugin  #
# --------------------------------------------- #
run-shell /nix/store/y1jjkdljiwlsqh52grr8xsqk41r3if96-tmuxplugin-sensible-unstable-2017-09-05/share/tmux-plugins/sensible/sensible.tmux
# ============================================= #

...

set  -g history-limit     2000

...
```

Which means that the sensible script is run first, which sets e.g. `history-limit` to 50.000, then the defaults from tmux kick in and reset history-limit back to 2000. This whole thing is just silly...

**Comment by ThomasRitaine** [comment]
For anyone hitting this issue, I've found a reliable workaround: place `sensible` at the **end** of your plugin list instead of using `sensibleOnTop`.

```nix
{ pkgs, ... }: {
  programs.tmux = {
    enable = true;
    # Other options
    sensibleOnTop = false;  # not mandatory as false is already the default

    plugins = with pkgs.tmuxPlugins; [
      # Your other plugins
      
      { plugin = sensible; }  # Load sensible last
    ];
  };
}
```
